src/compiler/checker.ts

```diff
@@ -280,6 +280,11 @@ namespace ts {
         const resolvingSignature = createSignature(undefined, undefined, undefined, emptyArray, anyType, /*typePredicate*/ undefined, 0, /*hasRestParameter*/ false, /*hasLiteralTypes*/ false);
         const silentNeverSignature = createSignature(undefined, undefined, undefined, emptyArray, silentNeverType, /*typePredicate*/ undefined, 0, /*hasRestParameter*/ false, /*hasLiteralTypes*/ false);
 
+        const unresolvedTypePredicate: void & { __unresolvedTypePredicate: void } = (() => {

```

Comment:
I would disagree that it's not an issue -- this was a barrier to my understanding this code, and it would have been easier had I realized that `resolvedReturnType` was not meant to be used directly as a `Type`. I hadn't come across `noConstraintType` yet, but I'm sure it wouldn't have been easy to realize that a value of type `Type` should not actually be used as a `Type` because it might be a special sentinel value.